### ISSUE TITLE
🚀 Feature: Expose granular API for consumers like eslint-plugin-package-json

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [JoshuaKGoldberg/package.json-validator#80]
## Context

I maintain [`eslint-plugin-package-json`](https://github.com/JoshuaKGoldberg/eslint-plugin-package-json), which is an [ESLint](https://eslint.org) plugin that lints for `package.json` file correctness. It includes `package.json-validator` by way of a [`package-json/valid-package-def` rule](https://github.com/JoshuaKGoldberg/eslint-plugin-package-json/blob/9077496bb1a4d4f63144b8a15759ca9b8c75593c/docs/rules/valid-package-def.md) that directly runs `package.json-validator`'s Node.js API and reports back any complaints.

## Problem Statement

That strategy technically works for `eslint-plugin-package-json`, but:

* Sometimes feature requests come into `package.json-validator` that happen to overlap other existing `eslint-plugin-package-json` rules
* It'd be difficult to add granular `package-json/valid-package-def` configuration: e.g. avoiding only some warnings
* It's confusing how a _lot_ of `eslint-plugin-package-json`'s functionality is encapsulated in one rule

Taking a step back, `package.json-validator` does more than just check for a _"valid"_ package file. It also gives warnings and recommendations that are more akin to linter warnings.

## Proposal

What if `package.json-validator` were to additionally expose individual API functions for each of its checks? Then consumers could use the individual API piece they want as needed.

In this case, `eslint-plugin-package-json` could make a single configurable rule for each of those checks. It'd still have a dependency on the `package-json-validator` package. We could then deduplicate code so that users can get the same recommendations and warnings either via ESLint plugin or this package.

**Comment by michaelfaith** [comment]
The way I kind of see this going is, when a new rule is implemented in `eslint-plugin-package-json` that would benefit from a subset of what's already in the `validate` functionality in this library, then break that piece out and expose it here as a dedicated function. i.e., having demand drive what the public API ultimately becomes?  

What might become a bit tricky in terms of avoiding breaking changes, is if the external demand needs a slightly more rigorous version of that functionality.  If we take the restrict version range example.  Right now the validation on dependency versions in this library is purely just testing (in a very relaxed way) if a version conforms to some acceptable format.  If the eslint rule needs something more configurable, like requiring that only `^` ranges are used on a certain pattern of dependencies (e.g. `@jest/*`), the new `isDependencyVersionValid` function we export here would need to support that level of configuration.  But if that function is _also_ powering the version validation in the legacy `validate` function, do we incorporate some default version of the more strict validation, or try and make it as relaxed as the old `validate` test was to prevent breaking changes?  Or do we not have `validate` use the new function at all, and leave it alone, with the goal of ultimately retiring it, once all of the more granular functions are provided?

Furthermore, as `validate` starts to become decomposed into its component parts, how do we prevent `eslint-plugin-package-json` from double reporting violations?  e.g. reporting for the restrict version granular rule as well as the old more broad rule that uses `validate`.  Just some things to consider.

**Comment by JoshuaKGoldberg** [comment]
> then break that piece out and expose it here as a dedicated function. i.e., having demand drive what the public API ultimately becomes?

👍 strong agree with this direction. Just confirming we're in sync, I think in general we're starting to see two discrete use cases for this package:

* 🍎 All-at-once (e.g. traditional Node.js API): running `validate()` to get a full list of reports
* 🍌 Granular (e.g. via individual lint rules): looking for a single area of reports

That lends itself well IMO to the API exporting:

* 🍎 `validate`: runs all the individual areas of reporting, internally running each of...
* 🍌 `validate*`: individual rules for each area of reporting, like `validName`, `validRepositoryDirectory`, etc.

Each lint rule would call to one of the 🍌 `validate*` functions.

Is that what you're envisioning @michaelfaith? If so, 👍 I think we're on the same page. 

> how do we prevent `eslint-plugin-package-json` from double reporting violations

How about we, in one single version:

1. Deprecate the 🍎 [`package-json/valid-package-def` rule](https://github.com/JoshuaKGoldberg/eslint-plugin-package-json/blob/9077496bb1a4d4f63144b8a15759ca9b8c75593c/docs/rules/valid-package-def.md) and remove it from the recommended config
2. Promote individual 🍌 rules for all the individual `validate*` functions in the recommended config

That way users of the recommended config will receive the same set of lint reports. And then any level of configurability could be present in the rule's options -> the individual function. 

The individual 🍌 rules can be added ahead of time and just not included in recommended yet. That way the final config swap change isn't a huge PR.

**Comment by michaelfaith** [comment]
> 🍎 validate: runs all the individual areas of reporting, internally running each of...

Yeah, but I think the one thing I wanted to highlight is that if we did do it this way, then we may end up making `validate` more strict than it currently is (because we want the component validations to be more strict in some way (like the version range example)) and introduce breaking changes to the old api.

**Comment by JoshuaKGoldberg** [comment]
I actually don't mind that so much 😄 benefit of being on a 0.x version... I'd expect that a lot of those stricter validations would come across as nice feature improvements for users.

For anything _really_ painful we can always add an option to the individual `validate*` function, and maybe also an option to `validate` itself.

🚀 accepting PRs!

**Comment by michaelfaith** [comment]
Does it make sense to leave this issue open, since it was more about agreeing on the strategy itself.  The actual work should probably be done under separate issues (one for each new validate function), e.g. https://github.com/JoshuaKGoldberg/package-json-validator/issues/292.  Unless you think there's value in keeping this open as inspiration for those.  Or, I guess those could be sub-issues of this one...

**Comment by JoshuaKGoldberg** [comment]
👍 agreed, this discussion & issue is resolved. If folks want to add more 🍌s then a new targeted issue would be the right one.
